src/System.Net.Security/src/System/Net/CertificateValidationPal.OSX.cs

```diff
@@ -0,0 +1,167 @@
+            return GetRemoteCertificate(securityContext, null);
+        }
+
+        internal static X509Certificate2 GetRemoteCertificate(
+            SafeDeleteContext securityContext,
+            out X509Certificate2Collection remoteCertificateStore)
+        {
+            if (securityContext == null)
+            {
+                remoteCertificateStore = null;
+                return null;
+            }
+
+            remoteCertificateStore = new X509Certificate2Collection();
+            return GetRemoteCertificate(securityContext, remoteCertificateStore);
+        }
+
+        private static X509Certificate2 GetRemoteCertificate(
+            SafeDeleteContext securityContext,
+            X509Certificate2Collection remoteCertificateStore)
+        {
+            if (securityContext == null)
+            {
+                return null;
+            }
+
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(securityContext);
+
+            SafeSslHandle sslContext = ((SafeDeleteSslContext)securityContext).SslContext;
+
+            if (sslContext == null)
+            {
+                return null;
+            }
+
+            X509Certificate2 result = null;
+
+            using (SafeX509ChainHandle chainHandle = Interop.AppleCrypto.SslCopyCertChain(sslContext))
+            {
+                long chainSize = Interop.AppleCrypto.X509ChainGetChainSize(chainHandle);
+
+                if (remoteCertificateStore != null)
+                {
+                    for (int i = 0; i < chainSize; i++)
+                    {
+                        IntPtr certHandle = Interop.AppleCrypto.X509ChainGetCertificateAtIndex(chainHandle, i);
+                        remoteCertificateStore.Add(new X509Certificate2(certHandle));
+                    }
+                }
+
+                // This will be a distinct object than remoteCertificateStore[0] (if applicable),
+                // to match what the Windows and Unix PALs do.
+                if (chainSize > 0)
+                {
+                    IntPtr certHandle = Interop.AppleCrypto.X509ChainGetCertificateAtIndex(chainHandle, 0);
+                    result = new X509Certificate2(certHandle);
+                }
+            }
+
+            if (NetEventSource.IsEnabled)
+            {
+                NetEventSource.Log.RemoteCertificate(result);
+                NetEventSource.Exit(securityContext, result);
+            }
+
+            return result;
+        }
+
+        //
+        // Used only by client SSL code, never returns null.
+        //
+        internal static string[] GetRequestCertificateAuthorities(SafeDeleteContext securityContext)
+        {
+            SafeSslHandle sslContext = ((SafeDeleteSslContext)securityContext).SslContext;
+
+            if (sslContext == null)
+            {
+                return Array.Empty<string>();
+            }
+
+            using (SafeCFArrayHandle dnArray = Interop.AppleCrypto.SslCopyCADistinguishedNames(sslContext))
```

Comment:
Not clear on how individual native array elements are cleanup up here (or whether they should be)